openlibrary/plugins/openlibrary/js/SearchBar.js

```diff
@@ -0,0 +1,300 @@
+import { debounce } from './nonjquery_utils.js';
+import * as SearchUtils from './SearchUtils';
+import { PersistentValue } from './SearchUtils';
+import $ from 'jquery';
+
+/** Mapping of search bar facets to search endpoints */
+const FACET_TO_ENDPOINT = {
+    'title': '/search',
+    'author': '/search/authors',
+    'lists': '/search/lists',
+    'subject': '/search/subjects',
+    'all': '/search',
+    'text': '/search/inside'
+};
+const DEFAULT_FACET = 'all';
+/** Functions that render autocomplete results */
+const RENDER_AUTOCOMPLETE_RESULT = {
+    ['/search'](work) {
+        const author_name = work.author_name ? work.author_name[0] : '';
+        return `
+            <li>
+                <a href="${work.key}">
+                    <img src="//covers.openlibrary.org/b/id/${work.cover_i}-S.jpg"/>
+                    <span class="book-desc">
+                        <div class="book-title">${work.title}</div> by <span class="book-author">${author_name}</span>
+                    </span>
+                </a>
+            </li>`;
+    },
+    ['/search/authors'](author) {
+        // Todo: default author img to: https://dev.openlibrary.org/images/icons/avatar_author-lg.png
+        return `
+            <li>
+                <a href="/authors/${author.key}">
+                    <img src="http://covers.openlibrary.org/a/olid/${author.key}-S.jpg"/>
+                    <span class="author-desc"><div class="author-name">${author.name}</div></span>
+                </a>
+            </li>`;
+    }
+}
+
+/**
+ * Manages the interactions associated with the search bar in the header
+ */
+export class SearchBar {
+    /**
+     * @param {HTMLElement|JQuery}
+     * @param {Object?} urlParams
+     */
+    constructor(component, urlParams={}) {
+        /** UI Elements */
+        this.$component = $(component);
+        this.$form = this.$component.find('form.search-bar-input');
+        this.$input = this.$form.find('input[type="text"]');
+        this.$results = this.$component.find('ul.search-results');
+        this.$facetSelect = this.$component.find('.search-facet-selector select');
+
+        /** State */
+        /** Whether the bar is in collapsible mode */
+        this.inCollapsibleMode = false;
+        /** Whether the search bar is currently collapsed */
+        this.collapsed = false;
+        /** Selected facet (persisted) */
+        this.facet = new PersistentValue('facet', {
+            default: DEFAULT_FACET,
+            initValidation(val) { return val in FACET_TO_ENDPOINT; }
+        });
+
+        this.initFromUrlParams(urlParams);
+        this.initCollapsibleMode();
+
+        // Bind to changes in the search state
+        SearchUtils.mode.sync(this.handleSearchModeChange.bind(this));
```

Comment:
I couldn't think of a better name :/ I opted for having something that looks slightly confusing so that someone will have to jump to the other class to read it, as opposed to them making assumptions if I use a more common name like `on` or `subscribe`